File: Ecommerce/backend/Ecommerce.BLL/ProductService.cs
using AutoMapper;
using Ecommerce.Contracts.Repositories;
using Ecommerce.Contracts.Services;
using Ecommerce.Models;
using Ecommerce.Models.DTOs;
using Ecommerce.Shared.Exceptions;

namespace Ecommerce.BLL
{
    public class ProductService : IProductService
    {
        private readonly IProductRepository _productRepository;
        private readonly IVendorRepository _vendorRepository;
        private readonly ICurrentUserService _currentUser;
        private readonly IMapper _mapper;

        public ProductService(IProductRepository productRepository, IVendorRepository vendorRepository, ICurrentUserService currentUser, IMapper mapper)
        {
            _productRepository = productRepository;
            _vendorRepository = vendorRepository;
            _currentUser = currentUser;
            _mapper = mapper;
        }

        public async Task<PageResponse<ProductResponse>> GetProductsCatalog(ProductFilterRequest query)
        {
            if (!string.IsNullOrEmpty(query.SortOrder) && !(query.SortOrder == "asc" || query.SortOrder == "desc"))
                throw new ValidationException("Sort Order must be 'asc' or 'desc'");
            if (!string.IsNullOrEmpty(query.SortBy) && !(query.SortBy == "price" || query.SortBy == "newest" || query.SortBy == "rating" || query.SortBy == "discount"))
                throw new ValidationException("Sort By must be 'price', 'newest', 'rating' or 'discount'");

            bool onlyDiscounted = string.Equals(query.SortBy, "discount", StringComparison.OrdinalIgnoreCase);

            var products = await _productRepository.GetProductsPaged(query.PageNumber, query.PageSize,query.SortBy,query.SortOrder,query.CategoryId,query.MinPrice,query.MaxPrice,query.SearchQuery);

            var totalCount = await _productRepository.GetProductsCount(query.CategoryId,query.MinPrice,query.MaxPrice,onlyDiscounted,query.SearchQuery);

            return new PageResponse<ProductResponse>
            {
                Items = _mapper.Map<ICollection<ProductResponse>>(products),
                PageNumber = query.PageNumber,
                PageSize = query.PageSize,
                TotalCount = totalCount
            };
        }

        public async Task<ICollection<ProductSearchResult>> SearchProducts(string searchTerm)
        {
            var products = await _productRepository.SearchProductsByName(searchTerm);
            return _mapper.Map<ICollection<ProductSearchResult>>(products);    
        }

        public async Task<ProductResponse> GetProductDetails(int productId)
        {
            var product = await _productRepository.GetById(productId) ?? throw new InvalidProductException($"Product with ID: {productId} not found.");
            return _mapper.Map<ProductResponse>(product);
        }

        public async Task<ICollection<ProductResponse>> GetVendorProducts()
        {
            var vendor = await _vendorRepository.GetByUserId(_currentUser.UserId) ?? throw new InvalidOwnershipException("Vendor account not found.");
            var products = await _productRepository.GetProductsByVendorId(vendor.Id);
            return _mapper.Map<ICollection<ProductResponse>>(products);
        }

        public async Task<ICollection<ProductResponse>> GetProductsByVendorId(int vendorId)
        {
            var products = await _productRepository.GetProductsByVendorId(vendorId);
            return _mapper.Map<ICollection<ProductResponse>>(products);
        }

        public async Task<ProductResponse> CreateProduct(CreateProductRequest request)
        {
            var vendorId = await GetCurrentVendorId();

            if (await _productRepository.Exists(p => p.VendorId == vendorId && p.Name.ToLower().Trim() == request.Name.ToLower().Trim()))
            {
                throw new UniquenessViolationException($"Product name '{request.Name}' already exists.");
            }

            var product = _mapper.Map<Product>(request);
            product.VendorId = vendorId;
            product.CreatedAt = DateTime.Now;
            product.Status = ProductStatus.Draft;

            var created = await _productRepository.Create(product);
            return _mapper.Map<ProductResponse>(created);
        }

        public async Task<ProductResponse> UpdateProduct(int productId, UpdateProductRequest request)
        {
            var product = await _productRepository.GetById(productId) ?? throw new InvalidProductException($"Product with ID {productId} not found.");

            await EnsureVendorOwns(product);

            if (!string.IsNullOrEmpty(request.Name) && !string.Equals(product.Name, request.Name.Trim(), StringComparison.OrdinalIgnoreCase))
            {
                if (await _productRepository.Exists(p => p.VendorId == product.VendorId && p.Id != productId && p.Name.ToLower().Trim() == request.Name.ToLower().Trim()))
                {
                    throw new UniquenessViolationException($"Product name '{request.Name}' already exists.");
                }
                product.Name = request.Name;
            }
            if (!string.IsNullOrEmpty(request.Description)) product.Description = request.Description;
            if (request.CategoryId.HasValue) product.CategoryId = request.CategoryId.Value;

            await _productRepository.Update(productId, product);
            return _mapper.Map<ProductResponse>(product);
        }

        public async Task<ProductResponse> ToggleProductStatus(int productId)
        {
            var product = await _productRepository.GetById(productId) ?? throw new InvalidProductException($"Product with ID {productId} not found.");

            await EnsureVendorOwns(product);
            if(product.Status == ProductStatus.Archived){
                ValidateProductForPublish(product);
                product.Status = ProductStatus.Active;
            }
            else {
                product.Status = ProductStatus.Archived;
                foreach(var variant in product.Variants)
                {
                    variant.IsActive = false;
                }
            }
            
            await _productRepository.Update(productId, product);
            return _mapper.Map<ProductResponse>(product);
        }

        
        public async Task<ProductResponse?> PublishProduct(int productId)
        {
            var product = await _productRepository.GetById(productId) ?? throw new InvalidProductException("Product not found.");
            await EnsureVendorOwns(product);
            ValidateProductForPublish(product);
            
            product.Status = ProductStatus.Active;
            var result = await _productRepository.Update(productId, product);
            return _mapper.Map<ProductResponse>(result);
        }

        private async Task EnsureVendorOwns(Product product)
        {
            var vendorId = await GetCurrentVendorId();
            if (product.VendorId != vendorId) throw new InvalidOwnershipException("Access denied! You do not own this product.");
        }

        private async Task<int> GetCurrentVendorId()
        {
            var vendor = await _vendorRepository.GetByUserId(_currentUser.UserId) ?? throw new InvalidOwnershipException("Vendor account not found.");
            return vendor.Id;
        }
        private void ValidateProductForPublish(Product product){
            bool hasValidActiveVariant = product.Variants.Any(v => v.IsActive && v.Price > 0 && v.StockQty >= 0);
            if (!hasValidActiveVariant) throw new InvalidOperationException("Product must contain at least one active variant with a valid price and stock.");
            
            bool activeVariantsHaveImages = product.Variants.Where(v => v.IsActive).All(v => v.VariantImages != null && v.VariantImages.Any());
            if (!activeVariantsHaveImages) throw new InvalidOperationException("All active variants must have at least one image before publishing.");
        }

    }
}

File: Ecommerce/frontend/src/app/services/product.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { environment } from "../../environments/environment";
import { Observable } from "rxjs";
import { ProductResponse, PageResponse, ProductFilterRequest, CreateProductRequest, UpdateProductRequest, AddProductVariantRequest, ProductVariantResponse, CreateProductImageRequest, ProductSearchResult, UpdateProductVariantRequest } from "../models/product.model";

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private baseUrl = `${environment.baseUrl}/Product`;

  constructor(private http: HttpClient) {}

  getCatalog(filter?: ProductFilterRequest): Observable<PageResponse<ProductResponse>> {
    let params = new HttpParams();
    if (filter) {
      if (filter.pageNumber !== undefined) params = params.set('pageNumber', filter.pageNumber.toString());
      if (filter.pageSize !== undefined) params = params.set('pageSize', filter.pageSize.toString());
      if (filter.categoryId !== undefined) params = params.set('categoryId', filter.categoryId.toString());
      params = filter.sortBy ? params.set('sortBy', filter.sortBy): params.set('sortBy', 'newest');
      params = filter.sortOrder ? params.set('sortOrder', filter.sortOrder) : params.set('sortOrder', 'desc');
      if (filter.searchQuery != undefined ) params = params.set('searchQuery', filter.searchQuery);
      if (filter.minPrice !== undefined) params = params.set('minPrice', filter.minPrice.toString());
      if (filter.maxPrice !== undefined) params = params.set('maxPrice', filter.maxPrice.toString());
    }
    return this.http.get<PageResponse<ProductResponse>>(this.baseUrl, { params });
  }

  search(q: string): Observable<ProductSearchResult[]> {
    return this.http.get<ProductSearchResult[]>(`${this.baseUrl}/search?q=${q}`);
  }

  getById(id: number): Observable<ProductResponse> {
    return this.http.get<ProductResponse>(`${this.baseUrl}/${id}`);
  }

  getVendorProducts(pageNumber: number = 1, pageSize: number = 5): Observable<PageResponse<ProductResponse>> {
    return this.http.get<PageResponse<ProductResponse>>(`${this.baseUrl}/vendor?pageNumber=${pageNumber}&pageSize=${pageSize}`);
  }

  getProductsByVendorId(vendorId: number, pageNumber: number = 1, pageSize: number = 5): Observable<PageResponse<ProductResponse>> {
    return this.http.get<PageResponse<ProductResponse>>(`${this.baseUrl}/vendor/${vendorId}?pageNumber=${pageNumber}&pageSize=${pageSize}`);
  }

  createProduct(request: CreateProductRequest): Observable<{ message: string; data: ProductResponse }> {
    return this.http.post<{ message: string; data: ProductResponse }>(this.baseUrl, request);
  }

  updateProduct(id: number, request: UpdateProductRequest): Observable<{ message: string; data: ProductResponse }> {
    return this.http.patch<{ message: string; data: ProductResponse }>(`${this.baseUrl}/${id}`, request);
  }

  publishProduct(productId: number): Observable<{ message: string; data: ProductResponse }> {
    return this.http.patch<{ message: string; data: ProductResponse }>(`${this.baseUrl}/publish/${productId}`, {});
  }

  toggleProductStatus(id: number): Observable<{ message: string; data: ProductResponse }> {
    return this.http.patch<{ message: string; data: ProductResponse }>(`${this.baseUrl}/toggle/${id}`, {});
  }

  addVariant(productId: number, request: AddProductVariantRequest): Observable<{ message: string; data: ProductVariantResponse }> {
    return this.http.post<{ message: string; data: ProductVariantResponse }>(`${environment.baseUrl}/ProductVariant/${productId}`, request);
  }

  updateVariant(variantId: number, request: UpdateProductVariantRequest): Observable<{ message: string; data: ProductVariantResponse }> {
    return this.http.put<{ message: string; data: ProductVariantResponse }>(`${environment.baseUrl}/ProductVariant/${variantId}`, request);
  }

  toggleVariantStatus(variantId: number): Observable<{ message: string; data: any }> {
    return this.http.patch<{ message: string; data: any }>(`${environment.baseUrl}/ProductVariant/${variantId}`,{});
  }

  addVariantImage(variantId: number, request: CreateProductImageRequest): Observable<{ message: string; data: any }> {
    return this.http.post<{ message: string; data: any }>(`${environment.baseUrl}/ProductVariant/image/${variantId}`, request);
  }

  deleteVariantImage(imageId: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${environment.baseUrl}/ProductVariant/image/${imageId}`);
  }

  uploadVariantImage(file: File, productName: string, variantNo: number, imageNo: number, variantId: number): Observable<{ imageUrl: string }> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('productName', productName);
    formData.append('variantNo', variantNo.toString());
    formData.append('imageNo', imageNo.toString());
    formData.append('variantId', variantId.toString());
    return this.http.post<{ imageUrl: string }>(`${environment.baseUrl}/Upload/variant-image`, formData);
  }
}